### ISSUE TITLE
Add unit tests for api-tools/parser/sectionParser and api-tools/parser/courseParser

### ISSUE THREAD

**Issue by jpahm** [UTDNebula/api-tools#43]
Add unit tests using [Go's built-in testing library](https://pkg.go.dev/testing). I would personally be in favor of the test files being within the same package as the files being tested so that non-exported methods can be tested more easily; make sure to follow the `_test.go` file naming convention detailed in the linked docs. **Be sure to seek out resources for details on how to write good unit tests if you are unsure.**

**Comment by SoggyRhino** [comment]
Can I work on this?

**Comment by jpahm** [comment]
> Can I work on this?

For sure! Let me know if you need further clarification. I have a basic unit testing example established already for the utilities.

**Comment by SoggyRhino** [comment]
I've made tests for most of the actual parsing (ie html into objects), but I had a few questions/issues:

I'm struggling with part that deals with the maps. There is alot of interdependence and late modification so it doesn't seem feasible to test each part in isolation. If you ignore the html parsing, the logic isn't all that complex so it might be worth just testing the parser as a whole. Just feed it some html and compare the resulting Courses.json, Proffessors.json and Sections.json to what is expected. 

The grade loading code uses file creating with relative paths. This isn't an issue normally but unit tests have a different working directory than main.go (ie /api-tools/parser/ vs /api-tools/) causing it to fail when creating log files. I haven't spent much time trying to find a work around but I doesn't look like something can be handled in the tests.

The schema objects are hard to compare, refeclt.DeepEqual doesn't work on the ones with times because times need to be compared a special way. Also even it worked it wouldn't provided very useful messages. Something like ... Fail expected {very long json} got {very long json}. Originally I was using a bunch of if statements (one for each field) but it I found a library [cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) and works well. It has a diff function that makes the errors easier to understand. 

Example of cmp  (I added some garbage to the expected)
```bash
=== RUN   TestGetCourse/case_001
    courseParser_test.go:23: Failed (-expected +got)
           schema.Course{
          	... // 1 ignored field
          	Subject_prefix: "ACCT",
          	Course_number:  "2301",
          	Title: strings.Join({
          		"Introductory Financial Accounting",
        - 		"asdasd",
          	}, ""),
          	Description:     "ACCT 2301 - Introductory Financial Accounting  (3 semester credi"...,
        - 	Enrollment_reqs: "asdasd",
        + 	Enrollment_reqs: "",
          	School:          "Naveen Jindal School of Management",
          	Credit_hours:    "3",
          	... // 13 identical fields
          }
```

I didn't want to add a dependency but otherwise each test would need essentially make a new Equal() implementation. Using cmp also made the code much shorter and easier to read. It also has the benefit of not breaking if the schema changes since it uses reflection (except if you rename fields that are handled specially like ignoring Id) .

 Its seems fine, ran by google, popular and mature. If you could look into it that would be great. Thanks

**Comment by jpahm** [comment]
Hi, thanks for looking so deeply into this!

I think the approach you mention with using cmp to test the parser as a whole is good -- I certainly don't mind adding a dependency if it makes something as crucial as testing easier with less maintenance overhead.

As far as the pathing issue with the grade loader, that's something we can adjust to be more unit testable. I could make a separate issue to redesign it if you'd like, otherwise you're welcome to take a crack at it as part of this one.

**Comment by SoggyRhino** [comment]
I think it should be a new issue. I've never written any unit tests before this so I want to get some feed back on this one before starting anything else. 

If its alright I'm going to make a pr with what I have. It covers section and course, excluding grade data and the overall parser. I'm not sure if its 100% but I don't have any frame of reference, I was just going off of what looks right.

**Comment by jpahm** [comment]
>If its alright I'm going to make a pr with what I have

That's absolutely fine, thanks! I'll review it and get you some feedback.

**Comment by jpahm** [comment]
Attaching this to #60.

**Comment by jpahm** [comment]
Completed via #60.
